lucene/core/src/java/org/apache/lucene/util/BytesRefArray.java

```diff
@@ -185,20 +186,48 @@ public BytesRefIterator iterator() {
    */
   @Override
   public BytesRefIterator iterator(final Comparator<BytesRef> comp) {
-    final BytesRefBuilder spare = new BytesRefBuilder();
-    final BytesRef result = new BytesRef();
+    return iteratorProvider(comp).get();
+  }
+
+  /**
+   * Prefer using either {@link #iterator()} or {@link #iterator(Comparator)}.
+   * This method is only useful if multiple iterators with a non-null {@link Comparator}
+   * are requires as it avoids sorting the array multiple times.
+   */
+  public Supplier<Iterator> iteratorProvider(final Comparator<BytesRef> comp) {
     final int size = size();
     final int[] indices = comp == null ? null : sort(comp);
-    return new BytesRefIterator() {
-      int pos = 0;
+
+    return () -> new Iterator() {
+      final BytesRefBuilder spare = new BytesRefBuilder();
+      final BytesRef result = new BytesRef();
+      int pos = -1;
+
       @Override
       public BytesRef next() {
+        ++pos;
         if (pos < size) {
-          setBytesRef(spare, result, indices == null ? pos++ : indices[pos++]);
+          setBytesRef(spare, result, currentIndex());
           return result;
         }
         return null;
       }
+
+      @Override
+      public int currentIndex() {
+        return indices == null ? pos : indices[pos];
+      }
     };
   }
+
+  /**
+   * An extension of {@link BytesRefIterator} that allows retrieving the index of the current element
+   */
+  public interface Iterator extends BytesRefIterator {
+    /**
+     * Returns the index of the element that was returned by the latest {@link #next()}. Do not call
+     * this method if {@link #next()} is not called yet or the last call returned a null value.
+     */
+    int currentIndex();
```

Comment:
ord() to be consistent to TermsEnum Contract? Additionally It would be nice to have this in a different class as today TermsEnum are used for DocValues forcing to un-implement several methods such as postings and TermStates

Reply:
I don't understand the second part of your comment about having this in a different class?

Reply:
@jpountz my mistake I wanted to say a different file like "SeekableBytesRefIterator" decoupling postings and terms iteration.

Reply:
I renamed it to `ord`